File: python/Backtracking/subsets_two.py
from typing import List

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        """
        Approach:
        - Sort the input to group duplicates together.
        - Use backtracking to generate all possible subsets.
        - At each index, decide to include or exclude the current element.
        - When excluding, skip all consecutive duplicates to avoid duplicate subsets.
        - When the end of the list is reached, add the current subset to the result.

        Time Complexity: O(2^n), where n is the length of nums (number of subsets).
        Space Complexity: O(2^n * n), for the result list (all subsets) and recursion stack.
        """
        res = []
        nums.sort()

        def backtrack(i, subset):
            if i == len(nums):
                res.append(subset[::])
                return
            
            #All subsets that include nums[i]
            subset.append(nums[i])
            backtrack(i + 1, subset)
            subset.pop()

            #All subsets that do not include nums[i]
            while i + 1 < len(nums) and nums[i] == nums[i + 1]:
                i += 1
            backtrack(i + 1, subset)
        backtrack(0, [])
        return res

# -------------------
# Test Cases
# -------------------
def test_subsets_with_dup():
    sol = Solution()
    # Test case 1: Example from Leetcode
    nums1 = [1,2,2]
    result1 = sol.subsetsWithDup(nums1)
    expected1 = [[],[1],[1,2],[1,2,2],[2],[2,2]]
    assert sorted([tuple(sorted(x)) for x in result1]) == sorted([tuple(sorted(x)) for x in expected1])
    # Test case 2: All unique
    nums2 = [1,2,3]
    result2 = sol.subsetsWithDup(nums2)
    expected2 = [[],[1],[2],[3],[1,2],[1,3],[2,3],[1,2,3]]
    assert sorted([tuple(sorted(x)) for x in result2]) == sorted([tuple(sorted(x)) for x in expected2])
    # Test case 3: All duplicates
    nums3 = [2,2,2]
    result3 = sol.subsetsWithDup(nums3)
    expected3 = [[],[2],[2,2],[2,2,2]]
    assert sorted([tuple(sorted(x)) for x in result3]) == sorted([tuple(sorted(x)) for x in expected3])
    # Test case 4: Empty list
    nums4 = []
    assert sol.subsetsWithDup(nums4) == [[]]
    print("All test cases passed!")

if __name__ == "__main__":
    test_subsets_with_dup()

File: C++/Backtracking/subsets_two.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Approach:
 * - Sort the input array so duplicates are adjacent.
 * - Use backtracking to explore all subsets.
 * - At each step, decide to include the current number or not.
 * - If we choose to skip a number, skip all its duplicates to avoid duplicate subsets.
 *
 * Time Complexity: O(2^n), because each element can be included or not.
 * Space Complexity: O(2^n * n), for storing all subsets and recursion stack.
 */

class Solution {
public:
    vector<vector<int>> res; // To store all unique subsets

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end()); // Sort to group duplicates together
        vector<int> subset;             // Current subset being built
        backtrack(0, nums, subset);
        return res;
    }

    void backtrack(int i, vector<int>& nums, vector<int>& subset) {
        if (i == nums.size()) {
            res.push_back(subset); // Base case: reached the end, add subset
            return;
        }

        // Include nums[i]
        subset.push_back(nums[i]);
        backtrack(i + 1, nums, subset);
        subset.pop_back(); // Backtrack

        // Skip duplicates
        while (i + 1 < nums.size() && nums[i] == nums[i + 1]) {
            i++;
        }

        // Exclude nums[i] and move to next distinct number
        backtrack(i + 1, nums, subset);
    }
};

/*
------------------------------------------------------
📄 Line-by-Line C++ Syntax Explanation:
------------------------------------------------------

#include <bits/stdc++.h>
    → Includes all standard C++ libraries (vector, algorithm, etc.) in one line.

using namespace std;
    → So we don’t need to prefix std:: before every type.

class Solution { ... };
    → Defines a class named Solution.

vector<vector<int>> res;
    → 2D vector to store the resulting subsets.

vector<vector<int>> subsetsWithDup(vector<int>& nums)
    → Public function that takes a reference to a vector and returns all unique subsets.

sort(nums.begin(), nums.end());
    → Sort the input array so duplicates are next to each other.

vector<int> subset;
    → Temporary vector to hold the current subset being constructed.

backtrack(0, nums, subset);
    → Start recursive backtracking at index 0.

void backtrack(int i, vector<int>& nums, vector<int>& subset)
    → Recursive function to generate subsets.

if (i == nums.size()) { ... }
    → Base case: if we've processed all elements.

res.push_back(subset);
    → Add the current subset to the result.

subset.push_back(nums[i]);
    → Include nums[i] in the current subset.

backtrack(i + 1, nums, subset);
    → Recur with the next index.

subset.pop_back();
    → Backtrack: remove last element to try other combinations.

while (i + 1 < nums.size() && nums[i] == nums[i + 1]) { i++; }
    → Skip over consecutive duplicates.

backtrack(i + 1, nums, subset);
    → Recur by skipping the current number.

return res;
    → Return all unique subsets after exploration.

------------------------------------------------------
⏱ Complexity:
- Time: O(2^n), where n is the number of elements.
- Space: O(2^n * n), for storing all subsets and recursion stack.
------------------------------------------------------
*/

